File: 3152-maximum-value-of-an-ordered-triplet-ii/3152-maximum-value-of-an-ordered-triplet-ii.c
#define max(a, b) (a > b) ? a : b;
long long maximumTripletValue(int *nums, int numsSize) {
  long long ans = 0;
  int max = 0, dmax = 0;
  for (int i = 0; i < numsSize; i++) {
    ans = max(((long long)dmax * nums[i]), ans);
    dmax = max(max - nums[i], dmax);
    max = max(max, nums[i]);
  }
  return ans;
}

File: 3152-maximum-value-of-an-ordered-triplet-ii/3152-maximum-value-of-an-ordered-triplet-ii.java
class Solution {
  public long maximumTripletValue(int[] nums) {
    long res = 0 ;
    int max = 0 , dmax = 0 ;
    for ( int num : nums )  {
      res = Math.max( res , (long) dmax * num ) ;
      dmax = Math.max(dmax, max-num) ;
      max = Math.max(max, num) ;
    }
    return res ;
  }
}
